File: C#/Pagination with JSON/Pagination.cs
using System;
using System.IO;
using System.Net;
using System.Collections.Generic;

// Newtonsoft.Json is available from http://www.newtonsoft.com/json
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;

public class BasicAuth
{
    static public void Main()
    {
        string baseURL = "{base url}/rest/latest/";

        // Username and password should be stored according
        // to your organization's security policies
        string username = "API_User";
        string password = "********";

        string resource = "projects";

        int allowedResults = 20;
        string maxResults = "maxResults=" + allowedResults;

        long resultCount = -1;
        long startIndex = 0;

        while (resultCount != 0)
        {
            string startAt = "startAt=" + startIndex;

            string url = baseURL + resource + "?" + startAt + "&" + maxResults;

            HttpWebRequest request = (HttpWebRequest)WebRequest.Create(url);
            request.Credentials = new NetworkCredential(username, password);
            request.Method = WebRequestMethods.Http.Get;
            request.PreAuthenticate = true;

            HttpWebResponse response = (HttpWebResponse)request.GetResponse();

            Stream stream = response.GetResponseStream();
            StreamReader streamReader = new StreamReader(stream);

            JObject jsonResponse = JObject.Parse(streamReader.ReadToEnd());
            JObject meta = (JObject) jsonResponse["meta"];
            JObject pageInfo = (JObject) meta["pageInfo"];

            startIndex = (long) pageInfo["startIndex"] + allowedResults;
            resultCount = (long) pageInfo["resultCount"];

            JArray projects = (JArray) jsonResponse["data"];
            foreach (JObject project in projects)
            {
                JObject fields = (JObject) project["fields"];
                string name = (string) fields["name"];
                Console.WriteLine(name);
            }
        }
    }
}

File: Ruby/Pagination with JSON/pagination.rb
require "net/http"
require "json"

base_url = "{base url}/rest/latest/"

# Username and password should be stored according
# to your organization's security policies
username = "API_User"
password = "********"

resource = "projects"

allowed_results = 20
max_results = "maxResults=#{allowed_results}"

result_count = -1
start_index = 0

while result_count != 0 do
    start_at = "startAt=#{start_index}"

    uri = URI("#{base_url}#{resource}?#{start_at}&#{max_results}")
    Net::HTTP.start(uri.host, uri.port, 
        :use_ssl => uri.scheme == 'https') do |http| 
        request = Net::HTTP::Get.new(uri)
        request.basic_auth(username, password)
        response = http.request(request)
        
        json_response = JSON.parse(response.body)

        page_info = json_response["meta"]["pageInfo"]
        start_index = page_info["startIndex"] + allowed_results
        result_count = page_info["resultCount"]

        projects = json_response["data"]
        projects.each do |project| 
            puts project["fields"]["name"]
        end
    end
end
